### PR TITLE
Bundle library using rollup

### PR DESCRIPTION
**Pull request by st3iny**
This library is now bundled using rollup. I also upgraded our npm lock file to version 2, removed some unnecessary dependencies and fixed all (2) linter errors.

### CODE REVIEW

**Comment by ChristophWurst** [comment]
How would I best test this? Are the imports identical for code that already used this package?

**Comment by st3iny** [comment]
> How would I best test this? Are the imports identical for code that already used this package?

Unfortunately not. I tried splitting the code similar to what we do in nextcloud-vue but that wasn't possible because there were too many side effects. Everything has to be imported from the root (`import { ... } from '@nextcloud/calendar-js'`) from now on.

I already refactored calendar locally, to be able to test my progress and I'll push the branch soon. That could be used for testing.

**Comment by ChristophWurst** [comment]
> Everything has to be imported from the root

I'm assuming this is okay because with the modern modules webpack and friend will be able to shake away the bits that are not used, right?

**Comment by st3iny** [comment]
> > Everything has to be imported from the root
> 
> I'm assuming this is okay because with the modern modules webpack and friend will be able to shake away the bits that are not used, right?

Yes that should be fine. I made use of the `module` field of package.json which may point to an esm file. Altough it still is a proposal, almost all bundlers accept it and actually prefer it over the default `main` field. E.g. Webpack will now read the `module` field and then load the esm file which enables tree shaking.

**Comment by ChristophWurst** [comment]
> that wasn't possible because there were too many side effects

Could you elaborate on that?

File: src/index.js [hunk 1]
```diff
@@ -19,17 +19,49 @@
...
+import { getTimezoneManager, TimezoneManager } from './timezones/timezoneManager.js'
 import { v4 as uuid } from 'uuid'
-import DateTimeValue from './values/dateTimeValue.js'
 import { dateFactory } from './factories/dateFactory.js'
-import CalendarComponent from './components/calendarComponent.js'
-import EventComponent from './components/root/eventComponent.js'
-import RecurrenceManager from './recurrence/recurrenceManager.js'
-import FreeBusyComponent from './components/root/freeBusyComponent.js'
 import ICAL from 'ical.js'
+import AbstractComponent from './components/abstractComponent'
```
Comment (tcitworld) [inline src/index.js:27]: All of these are not very readable. Why not import & export everything while grouping by folder, from `components/index.js`, `errors/index.js`, etc. instead?

Comment (st3iny) [reply src/index.js:27]: That is a good point. I'll refactor this.

**Comment by st3iny** [comment]
I added the whole src directory to the `files` field of package.json for now, to not break all apps depending on this library. This means that transpiling the source code locally or just using the bundled version will both work. In the future we could remove all source files from the package and bump the major version to indicate a breaking change.

**Comment by st3iny** [comment]
> > that wasn't possible because there were too many side effects
> 
> Could you elaborate on that?

While testing the calendar app, I encountered strange type errors or null vars that shouldn't be null. However, all those errors did not show up when I tested using the source code. My guess is that the bundler creates multiple version of each module in each separate bundle. The code is very interdependent and there are many imports from one subdir to the next. 

If we investigate and properly refactor the structure we could split the library. IMO, the potential gain isn't big enough right now. I already applied a big optimization by telling rollup to externalize ICAL so that it'll be bundled only once (by the app depending on calendar-js).

PS: I just checked and the bundle size of calendar shrinks by about ~10kb when using the bundled version.

**Comment by ChristophWurst** [comment]
> In the future we could remove all source files from the package and bump the major version to indicate a breaking change.

I'm also fine with doing this now. There are not many apps that depend on this package.

**Comment by st3iny** [comment]
> > In the future we could remove all source files from the package and bump the major version to indicate a breaking change.
> 
> I'm also fine with doing this now. There are not many apps that depend on this package.

Done. I also bumped the major version.

**Review by ChristophWurst — APPROVED** [review]
Looks good!

**Comment by ChristophWurst** [comment]
v2 released :shipit:

**Pull request merged**